src/s2/util/math/exactfloat/bignum.cc

```diff
@@ -0,0 +1,884 @@
+
+  // Core algorithm: Exponentiation by squaring.
+  Bignum result(1);
+  Bignum base = *this;  // A mutable copy of the base.
+  uint32_t upow = static_cast<uint32_t>(pow);
+
+  while (upow > 0) {
+    if (upow & 1) {  // If current exponent bit is 1, multiply into result.
+      result *= base;
+    }
+    base *= base;
+    upow >>= 1;
+  }
+
+  return result;
+}
+
+// Computes a + b + carry and updates the carry.
+inline Bigit AddBigit(Bigit a, Bigit b, Bigit* absl_nonnull carry) {
+// Compilers such as GCC and Clang are known to be terrible at generating good
+// code for long carry chains on Intel. Using the _addcarry_u64 intrinsic (which
+// maps to the add/adc instructions produces a tight series of add/adc/adc/adc
+// instructions, whereas writing the loop manually often generates many add/adc
+// pairs with spurious bit twiddling.
+//
+// Using the intrinsic here improves benchmarks by ~30% when summing larger
+// Bignums together.
+//
+// See this SO discussion for more information:
+//   https://stackoverflow.com/questions/33690791
+//
+// Godbolt link for comparison:
+//   https://godbolt.org/z/cGnnfMbMn  (no intrinsics)
+//   https://godbolt.org/z/jnM1Y3Tjs  (intrinsics)
+#ifdef __x86_64__
+  static_assert(sizeof(Bigit) == sizeof(unsigned long long));
+  Bigit out;
+  *carry =
+      _addcarry_u64(*carry, a, b, reinterpret_cast<unsigned long long*>(&out));
+  return out;
+#else
+  auto sum = absl::uint128(a) + b + *carry;
+  *carry = absl::Uint128High64(sum);
+  return static_cast<Bigit>(sum);
+#endif
+}
+
+// Computes a - b - borrow  and updates the borrow.
+//
+// NOTE: Borrow must be one or zero.
+inline Bigit SubBigit(Bigit a, Bigit b, Bigit* absl_nonnull borrow) {
+  ABSL_DCHECK_LE(*borrow, Bigit(1));
+  // See notes in AddBigit on why using an intrinsic is the right choice here.
+#ifdef __x86_64__
+  Bigit out;
+  *borrow = _subborrow_u64(*borrow, a, b,
+                           reinterpret_cast<unsigned long long*>(&out));
+  return out;
+#else
+  Bigit diff = a - b - *borrow;
+  *borrow = (a < b) || (*borrow && (a == b));
+  return diff;
+#endif
+}
+
+// Computes a * b + carry and updates the carry.
+inline Bigit MulBigit(Bigit a, Bigit b, Bigit* absl_nonnull carry) {
+  auto sum = absl::uint128(a) * b + *carry;
+  *carry = absl::Uint128High64(sum);
+  return static_cast<Bigit>(sum);
+}
+
+// Computes sum += a * b + carry and updates the carry.
+//
+// NOTE: Will not overflow even if a, b, and c are their maximum values.
+inline void MulAddBigit(Bigit* absl_nonnull sum, Bigit a, Bigit b,
+                        Bigit* absl_nonnull carry) {
+  // Similar to the comment in AddBigit, and just for completeness, it's worth
+  // noting that the "best" way to implement this is with the Intel MULX, ADCQ,
+  // and ADOQ instructions (i.e. the _mulx_u64, _addcarry_u64, and
+  // _addcarryx_u64 intrinsics), but GCC and Clang do not support _addcarryx_u64
+  // properly (and have no plans to do so). The issue is that gcc doesn't
```

Comment:
Do you have a link for "have no plans to do so"?

Reply:
These can be found with AI search mode. Since I want to wait a bit longer for Eric, you can add them. Otherwise, I can add links later.